### PR TITLE
Run TestVersionUpgradeOrdering with/without Logstash

### LINKED ISSUE #6928: TestVersionUpgradeOrdering is always skipped since Logstash was added
`TestVersionUpgradeOrdering` is not run since it has been updated to test an upgrade of Logstash.

It is because the initial version used for the upgrade is too low to test Logstash.

https://github.com/elastic/cloud-on-k8s/blob/8e3beaf80aae6dec61e64371778b1d082f0dddc4/test/e2e/stack_test.go#L38-L40

The `SkipTest()` func of the Logstash builder returns `true` when the version is less than `8.6.0`, which is the case here as the LatestReleasedVersion7x is `7.17.8`.

https://github.com/elastic/cloud-on-k8s/blob/8e3beaf80aae6dec61e64371778b1d082f0dddc4/test/e2e/test/logstash/builder.go#L228-L232

https://github.com/elastic/cloud-on-k8s/blob/8e3beaf80aae6dec61e64371778b1d082f0dddc4/pkg/controller/common/version/version.go#L36

### PR DESCRIPTION
**Pull request by thbkrkr**
This fixes an issue where `TestVersionUpgradeOrdering` is not run since it has been updated
to test an upgrade with Logstash because the upgrade is from `7x` to `8x` and Logstash is 
supported by ECK from `8.6.0` by:
- removing Logstash from `TestVersionUpgradeOrdering`
- adding a new test `TestVersionUpgradeOrderingWithLogstash` that tests an upgrade from `8.6.0` to `8x`

This also:
- adds a call to `association.AllowVersion` at the start of the logstash reconciliation to prevent upgrading Logstash until Elasticsearch is upgraded
- configures an associated Elasticsearch to Logstash otherwise there is no reason not to wait to upgrade Logstash
- updates the `StackResourceVersions` methods that ensure a strict ordering is respected during the version upgrade

Resolves #6928.

### CODE REVIEW

**Comment by thbkrkr** [comment]
buildkite test this -f p=gke,t=TestVersionUpgradeOrdering

**Comment by thbkrkr** [comment]
buildkite test this -f p=gke,t=TestVersionUpgradeOrder

**Review by pebrc — APPROVED** [review]
LGTM. Should we also update the `IsValid` function to check for Logstash if it is defined?

**Comment by thbkrkr** [comment]
> Should we also update the IsValid function to check for Logstash if it is defined?

Yes. I also added the `AllowVersion` check in the Logstash controller as I think this are the missing bits to ensure ES and Logstash are upgraded in the right order.

**Comment by thbkrkr** [comment]
buildkite test this -f p=gke,t=TestVersionUpgradeOrder

**Comment by thbkrkr** [comment]
I found the bug: Logstash has not associated ES, so there is no reason not to wait to upgrade Logstash.

**Comment by pebrc** [comment]
> I found the bug: Logstash has not associated ES, so there is no reason not to wait to upgrade Logstash.

Ah 💥 probably worth a comment in that method!

File: test/e2e/stack_test.go [hunk 1]
```diff
@@ -29,65 +29,107 @@ import (
...
+
+	ls := logstash.NewBuilder("ls").WithVersion(initialVersion).
+		WithElasticsearchRefs(
+			// associate logstash to the es ref stored in the stackVersions
+			logstashv1alpha1.ElasticsearchCluster{
+				ObjectSelector: commonv1.ObjectSelector{
+					Namespace: stackVersions.Elasticsearch.ref.Namespace,
+					Name:      stackVersions.Elasticsearch.ref.Name,
+				},
+				ClusterName: "es",
```
Comment (thbkrkr) [inline test/e2e/stack_test.go:111]: @pebrc Would you say more than that?

**Comment by thbkrkr** [comment]
buildkite test this -f p=gke,t=TestVersionUpgradeOrder

File: test/e2e/stack_test.go (hunk 1 above)
Comment (pebrc) [reply test/e2e/stack_test.go:111]: That's even better than what I had in mind. I thought we just document that we don't have an association. But you are right why not just create the association.

**Review by pebrc — APPROVED** [review]
LGTM

**Pull request merged**